### ISSUE TITLE
Improve pod install latency

### ISSUE THREAD

**Issue by xster** [flutter/flutter#10873]
Make sure no redundant work is done with CocoaPod during builds. 

Make sure repeated builds has a fast pod step.

**Comment by xster** [comment]
Related nature of issue #7673

**Comment by mit-mit** [comment]
Tagging customer-blocker and posse per discussion with Al.

**Comment by xster** [comment]
@alardizabal hoping to bug you guys with some extra info. Seems like you were hitting builds where the `pod install` step was taking 30s. Was it reproducible constantly or can you describe the project setup a bit if it is reproducible?

**Comment by mit-mit** [comment]
I tried to repro this by creating an app with lots of plugins in it (see `pubspec.yaml` below). This was slow on the first `flutter run`:

```
mit-macbookpro2:plugintest mit$ flutter run
Launching lib/main.dart on iPhone 6 in debug mode...
Running pod install...                               20.8s
Running Xcode build...                               22.4s
```

But subsequent runs were fairly fast:

```
mit-macbookpro2:plugintest mit$ flutter run
Launching lib/main.dart on iPhone 6 in debug mode...
Running pod install...                                3.0s
Running Xcode build...                               15.8s
```

My `pubspec.yaml`:

```
name: plugintest
description: A new Flutter project.

dependencies:
  flutter:
    sdk: flutter
  sqflite: "^0.1.0"
  android_intent:
  battery:
  connectivity:
  google_sign_in:
  image_picker:
  package_info:
  path_provider:
  quick_actions:
  sensors:
  share:
  shared_preferences:
  url_launcher:
  firebase_analytics:
  firebase_auth:
  firebase_database:
  firebase_messaging:
  firebase_storage:
```

**Comment by alardizabal** [comment]
@xster @mit-mit Thanks for following up. I recently uninstalled Xcode 9 because I believe that was part of the slowdown. I'm not sure if you want to test building with Xcode 9. Since reverting back to Xcode 8.3.3 my build times have been more reasonable. My pubspec is below and for pods, I'm bringing in Firebase Auth and Stripe. My `flutter doctor` is below. The pod installs do get progressively quicker on subsequent builds. However, I'm still seeing a lot of observatory timeouts (see below).

```
dependencies:
  flutter:
    sdk: flutter
  path_provider: "^0.2.1+1"
  fluro: "^1.1.0"
  firebase_analytics: "^0.0.3"
  firebase_database: "^0.0.11"
  url_launcher: "^0.4.2+1"
  permit:
    path: ./plugins/permit
```

```
☆ flutter doctor
[✓] Flutter (on Mac OS X 10.12.5 16F73, locale en-US, channel master)
    • Flutter at /Users/albertlardizabal/dev/flutter
    • Framework revision cb4a54db8f (2 days ago), 2017-06-27 16:19:59 -0700
    • Engine revision 009d470d50
    • Tools Dart version 1.24.2

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.0)
    • Android SDK at /Users/albertlardizabal/Library/Android/sdk
    • Platform android-26, build-tools 26.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[-] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b, 
    ✗ libimobiledevice and ideviceinstaller are not installed or require updating. To update, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.0
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1.4)
    • Flutter plugin version 14.0
    • Dart plugin version 171.4694.29

[✓] Connected devices
    • Pixel                        • FA71A0302167                             • android-arm • Android 7.1.2 (API 25)
    • XT1032                       • TA9290K3WW                               • android-arm • Android 5.1 (API 22)
    • Albert Lardizabal's iPhone 6 • c0e109b7e55a826627aa0f3e79af8931534f3554 • ios         • iOS 11.0
    • Posse iPhone                 • ac24c3de3cbe39efed542095769ce83fc89630f7 • ios         • iOS 10.3.1
    • iPhone SE                    • FF1E1EF8-E9B0-48A1-97FD-CD606A0F586F     • ios         • iOS 10.3 (simulator)

[10:48:47] (al-stripe_test)
☆ flutter run -d ac24c3de3cbe39efed542095769ce83fc89630f7
Launching lib/main.dart on Posse iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: 3MT3K2SKK7
Running pod install...                               16.7s
Running Xcode build...                               46.7s
Timeout while attempting to retrieve URL for Observatory

[10:51:08] (al-stripe_test)
☆ flutter run -d ac24c3de3cbe39efed542095769ce83fc89630f7
Launching lib/main.dart on Posse iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: 3MT3K2SKK7
Running pod install...                               10.4s
Running Xcode build...                               45.9s
Timeout while attempting to retrieve URL for Observatory

[10:53:35] (al-stripe_test)
☆ flutter run -d ac24c3de3cbe39efed542095769ce83fc89630f7
Launching lib/main.dart on Posse iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: 3MT3K2SKK7
Running pod install...                                8.5s
Running Xcode build...                               62.9s
```

**Comment by xster** [comment]
@mit-mit, I tried your test case. 
https://gist.github.com/xster/c06305938e87dcbebac2918c3d2f9779

My outcomes are fairly similar.
Observations:

1.  Took 20s for the first run too. Though nothing in there looked 'inefficient'. The time was mostly spent git cloning various plugins.
2. The second run was faster. The only thing that stood out was that in repeated runs, the cost of `   1.4760s   -> Fetching podspec for 'url_launcher' from '~/.pub-cache/hosted/pub.dartlang.org/url_launcher-0.4.2+1/ios'` continues to be orders of magnitudes higher than other dependencies. Maybe something wrong there. 

@alardizabal, I couldn't reproduce your case of consistent slow pods in repeated executions but I don't have ./plugins/permit either. 

Perhaps you can help us there by going to your project's ios/ folder and running `pod install --verbose | gnomon`. You'd need to `npm install -g gnomon` first (though it's an awesome tool) and set your FLUTTER_FRAMEWORK_DIR to something like `path-to-flutter/bin/cache/artifacts/engine/ios`

**Comment by alardizabal** [comment]
@xster @mit-mit 

Seems like url_launcher takes a few seconds (you mentioned in your observations). The below is my first try. Subsequent tries were around 5-6 seconds total with 3-4 seconds for url launcher.

```☆ pod install --verbose | gnomon
   1.5209s     Preparing
   0.0035s   
   0.0001s   Analyzing dependencies
   0.0000s   
   0.0000s   Inspecting targets to integrate
   0.0001s     Using `ARCHS` setting to build architectures of target `Pods-Runn
             er`: (`arm64`)
   0.0001s   
   0.0001s   Finding Podfile changes
   0.0000s     M Flutter
   0.0000s     - Firebase
   0.0000s     - Stripe
   0.0000s     - firebase_analytics
   0.0000s     - firebase_database
   0.0000s     - path_provider
   0.0000s     - permit
   0.0000s     - url_launcher
   0.0000s   
   0.0000s   Fetching external sources
   0.0284s   -> Fetching podspec for `Flutter` from `/Users/albertlardizabal/dev
             /flutter/bin/cache/artifacts/engine/ios`
   0.0427s   -> Fetching podspec for `firebase_analytics` from `/Users/albertlar
             dizabal/.pub-cache/hosted/pub.dartlang.org/firebase_analytics-0.0.3
             /ios`
   0.0137s   -> Fetching podspec for `firebase_database` from `/Users/albertlard
             izabal/.pub-cache/hosted/pub.dartlang.org/firebase_database-0.0.11/
             ios`
   0.0096s   -> Fetching podspec for `path_provider` from `/Users/albertlardizab
             al/.pub-cache/hosted/pub.dartlang.org/path_provider-0.2.1+1/ios`
   0.0084s   -> Fetching podspec for `permit` from `/Users/albertlardizabal/Poss
             e/clients/project-name/project-name-flutter/plugins/permit/ios`
   5.1862s   -> Fetching podspec for `url_launcher` from `/Users/albertlardizaba
             l/.pub-cache/hosted/pub.dartlang.org/url_launcher-0.4.2+1/ios`
   0.0001s   
   0.0001s   Resolving dependencies of `Podfile`
   0.0000s   
   0.0000s   Comparing resolved specification to the sandbox manifest
   0.0001s     - Firebase
   0.0000s     - FirebaseAnalytics
   0.0000s     - FirebaseAuth
   0.0000s     - FirebaseCore
   0.0002s     - FirebaseDatabase
   0.0000s     - FirebaseInstanceID
   0.0000s     - Flutter
   0.0001s     - GTMSessionFetcher
   0.0000s     - GoogleToolboxForMac
   0.0000s     - Stripe
   0.0001s     - firebase_analytics
   0.0001s     - firebase_database
   0.0003s     - path_provider
   0.0001s     - permit
   0.0000s     - url_launcher
   0.0000s   
   0.0000s   Downloading dependencies
   0.0000s   
   0.0000s   -> Using Firebase (3.15.0)
   0.0001s   
   0.0000s   -> Using FirebaseAnalytics (3.7.0)
   0.0000s   
   0.0000s   -> Using FirebaseAuth (3.1.1)
   0.0000s   
   0.0000s   -> Using FirebaseCore (3.5.2)
   0.0000s   
   0.0000s   -> Using FirebaseDatabase (3.1.2)
   0.0000s   
   0.0000s   -> Using FirebaseInstanceID (1.0.9)
   0.0001s   
   0.0000s   -> Using Flutter (1.0.0)
   0.0000s   
   0.0001s   -> Using GTMSessionFetcher (1.1.9)
   0.0002s   
   0.0000s   -> Using GoogleToolboxForMac (2.1.1)
   0.0000s   
   0.0000s   -> Using Stripe (10.0.1)
   0.0000s   
   0.0000s   -> Using firebase_analytics (0.0.1)
   0.0000s   
   0.0000s   -> Using firebase_database (0.0.1)
   0.0000s   
   0.0000s   -> Using path_provider (0.0.1)
   0.0002s   
   0.0000s   -> Using permit (0.0.1)
   0.0000s   
   0.0001s   -> Using url_launcher (0.0.1)
   0.0000s     - Running pre install hooks
   0.0000s   
   0.0000s   Generating Pods project
   0.0000s     - Creating Pods project
   0.0000s     - Adding source files to Pods project
   0.0000s     - Adding frameworks to Pods project
   0.0000s     - Adding libraries to Pods project
   0.0000s     - Adding resources to Pods project
   0.0000s     - Linking headers
   0.0000s     - Installing targets
   0.0000s       - Installing target `GTMSessionFetcher` iOS 7.0
   0.0000s       - Installing target `GoogleToolboxForMac` iOS 5.0
   0.0000s       - Installing target `Stripe` iOS 8.0
   0.0000s       - Installing target `firebase_analytics` iOS 6.0
   0.0000s       - Installing target `firebase_database` iOS 6.0
   0.0000s       - Installing target `path_provider` iOS 8.0
   0.0000s       - Installing target `permit` iOS 8.0
   0.0000s       - Installing target `url_launcher` iOS 8.0
   0.0000s       - Installing target `Pods-Runner` iOS 8.0
   0.0000s     - Running post install hooks
   0.0000s       - Podfile
   0.0004s     - Writing Xcode project file to `Pods/Pods.xcodeproj`
   0.0001s       - Generating deterministic UUIDs
   0.0001s     - Writing Lockfile in `Podfile.lock`
   0.0002s     - Writing Manifest in `Pods/Manifest.lock`
   0.0001s   
   0.0001s   Integrating client project
   0.0000s   
   0.0000s   Integrating target `Pods-Runner` (`Runner.xcodeproj` project)
   0.0000s     - Running post install hooks
   2.7142s       - cocoapods-stats from `/Users/albertlardizabal/.rvm/gems/ruby-
             2.1.3/gems/cocoapods-stats-1.0.0/lib/cocoapods_plugin.rb`
   0.0001s   
   0.0000s   Sending stats
   0.0000s         - Firebase, 3.15.0
   0.0000s         - FirebaseAnalytics, 3.7.0
   0.0000s         - FirebaseAuth, 3.1.1
   0.0000s         - FirebaseCore, 3.5.2
   0.0000s         - FirebaseDatabase, 3.1.2
   0.0000s         - FirebaseInstanceID, 1.0.9
   0.0000s         - GTMSessionFetcher, 1.1.9
   0.0002s         - GoogleToolboxForMac, 2.1.1
   0.0577s         - Stripe, 10.0.1
   0.0002s   
             -> Pod installation complete! There are 8 dependencies from the Pod
             file and 15 total pods installed.

   0.2894s   -> Pod installation complete! There are 8 dependencies from the Podse no platform was specified. Please specify a platform for this target in your Podfile. See `https://guides.co             file and 15 total pods installed.
   0.0004s   
             
     Total   9.8943s
```

**Comment by xster** [comment]
Thanks for the extra info Albert! Seems like the url_launcher issue is a clearly actionable low-hanging fruit. Will focus on that part first.

**Comment by xster** [comment]
Learned some ruby and got a few more findings :)

TL;DR none of the pods were actually problematic. 

Added some performance tracing code in the cocoapods source (it was fairly easy to set up https://guides.cocoapods.org/contributing/dev-environment.html). Didn't see much problems in the [path source fetching](https://github.com/CocoaPods/CocoaPods/blob/master/lib/cocoapods/external_sources/path_source.rb#L9) or the [dependency resolution](https://github.com/CocoaPods/CocoaPods/blob/master/lib/cocoapods/installer/analyzer.rb#L742) parts of the code. 

Ruby also doesn't seem to flush stdout immediately which caused some confusion initially. In sync stdout mode, it turns out the only problem with repeated builds is funny enough its stats collecting component https://github.com/CocoaPods/cocoapods-stats which sends stats synchronously during commands. It accounted for ~50% of total pod runtime in my case running @mit-mit's test pubspec. Would probably be worse on slower networks.  

Will disable.

**Comment by mit-mit** [comment]
O M G.

Thanks for the hard digging @xster !
